src/instructlab/config/init.py

```diff
@@ -183,6 +208,114 @@ def init(
     )
 
 
+def hw_auto_detect() -> tuple[str | None, int | None, _train, bool]:
+    # Third Party
+    import torch
+
+    gpus = 0
+    total_vram = 0
+    train = _train()
+    chosen_vram = None
+    chosen_profile_name = None
+    gpu_name = ""
+    edited_cfg = False
+    # try nvidia
+    if torch.cuda.is_available():
+        click.echo("Detecting Hardware...")
+        gpus = torch.cuda.device_count()
+        for i in range(gpus):
+            properties = torch.cuda.get_device_properties(i)
+            gpu_name = properties.name
+            total_vram += properties.total_memory  # memory in B
+    vram = int(floor(convert_bytes_to_proper_mag(total_vram)[0]))
+    if vram == 0:
+        return None, None, train, False
+    matched, chosen_vram, chosen_profile_name, train = lookup_card(
+        gpu_name=gpu_name, gpu_count=gpus, vram=vram
+    )
+    # if we used the vRAM to determine, we need to tell the user that
+    if not matched:
+        # edit nproc per node in the case that we didnt match an exact profile, so we use all GPUs on our system
+        train.nproc_per_node = gpus
+        edited_cfg = True
+    return (chosen_profile_name, chosen_vram, train, edited_cfg)
+
+
+# lookup_card checks if the user's GPU exists in our supported profile map. If not, we choose one based off of their vRAM
+def lookup_card(
+    gpu_name, gpu_count, vram
+) -> tuple[bool, int | None, str | None, _train]:
+    profiles = PROFILE_MAPPINGS
+    card_entry = profiles.get(gpu_name)
+    chosen_profile_name = None
+    if card_entry is not None:
+        chosen_profile_name = f"Nvidia {gpu_count}x {gpu_name}"
+        # it is not guaranteed we match, even if we find an entry
+        # what if the vram isn't found or what if the gpu count doesn't match
+        matched, train = match_profile_based_on_gpu_count(
+            card_entry=card_entry, gpu_count=gpu_count, vram=vram
+        )
+        if matched:
+            return True, vram, chosen_profile_name, train
+    vram, chosen_profile_name, train = match_profile_based_on_vram(vram=vram)
+    # return False with the results of the vRAM matching to indicate we didn't find an exact match for this GPU configuration in our supported profiles
+    return False, vram, chosen_profile_name, train
+
+
+# match_profile_based_on_vram chooses a training profile based off the amount of vram in each profile
+# this profile will be modified to match the number of GPUs on the user's system
+def match_profile_based_on_vram(vram) -> tuple[int | None, str | None, _train]:
+    chosen_profile_name = None
+    chosen_vram = None
+    train = _train()
+    for group_gpu_name, list_of_count_vram_and_config in PROFILE_MAPPINGS.items():
+        for gpu_count_configs in list_of_count_vram_and_config:
+            if (
+                not isinstance(gpu_count_configs["gpu_count"], int)
```

Comment:
I assume the isinstance is intended to check if it exists right? If so there will be a keyerror here when it doesnt exist which looks like the code block isnt expecting as an outcome.

Reply:
this is more for a mypy check since it didn't like us using the fields before checking the type.

I guess if someone messes with the struct this will error out, i can probably handle this more elegantly. The Pr merged because you were the last requested reviewer, but I will follow up with a PR to make this better.